message_app_client/src/components/conversation.js

```diff
@@ -0,0 +1,99 @@
+import React from 'react';
+
+class ConversationCard extends React.Component {
+    constructor(props){
+        super(props);
+        this.state = {
+            lastMessageContent: null,
+            conversationStatus: null,
+            title: null,
+            conversationID: null,
+        }
+    }
+
+    componentDidMount(){
+        this.setState({
+            lastMessageContent: this.props.lastMessageContent,
+            conversationStatus: this.props.conversationStatus,
+            title: this.props.title,
+            conversationID: this.props.conversationID
+        });
+    }
+
+    componentDidUpdate(prevProps){
+        // Update the state when the conversation has a new lastMessage
+        if (this.props.lastMessageContent !== prevProps.lastMessageContent){
+            this.setState({
+                lastMessageContent: this.props.lastMessageContent
+            });
+        }
+
+        // Update the state when the conversation has a new status
+        if (this.props.conversationStatus !== prevProps.conversationStatus){
+            this.setState({
+                conversationStatus: this.props.conversationStatus
+            });
+        }
+    }
+
+    render() {
+        const title = this.state.title;
+        const conversationStatus = this.state.conversationStatus;
+        const conversationID = this.state.conversationID;
+
+        return (
+            <div className="conversation-card__title" conversation_id={conversationID} conversation_status={conversationStatus}>
+                {title} - {conversationStatus}
+            </div>
+        )
+    }
+}
+
+class ConversationContainer extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {};
+    }
+
+    componentDidMount() {
+        for (let i=0;i<this.props.conversations.length;i++){
+            let conversation = this.props.conversations[i];
+            
+            this.setState({
+                [conversation.title]: conversation
```

Comment:
This looks kinda wrong to me. Are you trying to get all the conversations in a list and save it?

Reply:
Yes, I'm trying to save all the conversations. However, I didn't save it as a list but as `key: value`. The key is the conversation's title and the value is the object containing all of that conversation's data. The reason is for the updating purpose. For example, when a user logged in, the client receives a list of 10 conversations (let's say conversation_1 to conversation_10) from the server.  Then, that user receives a new message in conversation_1, which means the server will send new data that contains the changes of that conversation_1 only (change in the `lastMessageContent` for example). Using setState will change the data of that conversation_1 without affecting other conversations. Without saving all the conversations in the state, I think the client will delete all the old ones and only render the new conversation (I haven't tested this yet. I will do it tmr) when the client receives a new update. I know saving all the conversations is very costly, which is why I make this PR to hear more about other opinions.

Reply:
In this case, I would use conversation ID instead of conversation title.

For a 2-person conversation, the title will probably be the username of the other participant, such as `user_1`
For a 3+ person conversation, the title will be `user1, user2,...`.

The thing is, there is no guarantee that the title will be unique, so using ID would be better, right?

Reply:
Yes, using ID is better. However that is not my point here.

What I'm trying to do is to use the render to change the display of a conversation when the client receives new updates. Till now, all I found is to store the values of all conversations to the state, so when the client receives new updates, the component will update the state, which leads to the render also being updated.

We can always get the conversationID and update it outside of the ConversationContainer component (which is what I did in the old version). However, I still want to find a way to update the conversations inside of the component because it is less confusing for other people to read the code. I hope this make sense ._.

Reply:
Two points here:
- I got your idea of what you are trying to achieve here and agree with it
- I have a suggestion to change the `key` of the component's `state` to `conversation ID` instead of `conversation title`